src/coreclr/jit/compiler.cpp

```diff
@@ -5287,12 +5295,20 @@ void Compiler::placeLoopAlignInstructions()
             bbHavingAlign->bbFlags |= BBF_HAS_ALIGN;
             minBlockSoFar = BB_MAX_WEIGHT;
             bbHavingAlign = nullptr;
+            currentLoopNum = block->bbNext->bbNatLoopNum;
 
             if (--loopsToProcess == 0)
             {
                 break;
             }
         }
+
+        if ((block->bbJumpDest != nullptr) && (block->bbJumpDest->isLoopAlign()))
+        {
+            // This is the last block of current loop. Reset the latestLoopNum so we can resume
+            // placing the align behind jmp for subsequent loops.
+            currentLoopNum = BasicBlock::NOT_IN_LOOP;
+        }
```

Comment:
The current implementation will make sure we reset the `currentLoopNum` for the earliest back-edge we see. With your suggestion, if we have multiple back-edges to the loop, and in between, one of the block ended with `jmp`, we won't be able to use that for hiding the alignment because the `currentLoopNum` will still match the `block->bbNatLoopNum`.

Reply:
That's correct, but do you think that's a limitation? After all, if the bbNatLoopNum still matches, it's still part of the same inner loop. Is this concern because the first back-edge is considered the loop end in `emitter::getLoopSize()`? I would argue that is incorrect, even if it's not an unreasonable simplification. I don't think this situation arises often anyway, as most loops have a single back edge to the top.

Reply:
Is this concern because the first back-edge is considered the loop end in `emitter::getLoopSize()`?

Yes. My thinking behind the existing logic is that if `getLoopSize()` only considers the first back-edge, then we should open the opportunity to find `jmp` after that point and use them if needed instead of waiting to complete the loop.

Reply:
I would argue, alternatively, that we should think of ways to fix `getLoopSize()` to respect the loop package's determination about what constitutes the inner loop.

Note that, even as is, there are a few other concerns: (1) bbJumpDest is part of a `union` and not valid for all jump types, (2) you don't check that the branch is to the beginning of the loop we care about; it could be a forward branch so a subsequent loop that is aligned.

Reply:
> I would argue, alternatively, that we should think of ways to fix getLoopSize() to respect the loop package's determination about what constitutes the inner loop.


I think that needs separate discussion and should be not part of this PR. The initial decision was to just check the earliest back-edge to determine the loop size. If we want something smarter, we can definitely look at the weights of other blocks (the ones that follows after the earlier back-edge) to determine if those blocks are hot enough and do we need to include them as part of our loop size. 

> (1) bbJumpDest is part of a union and not valid for all jump types, 

That's true. We [check that ](https://github.com/dotnet/runtime/blob/ad9979381706d1976e4d26e13cb88d6147264968/src/coreclr/jit/codegenlinear.cpp#L755-L775) in `codegenlinear.cpp` and I should add similar check here as well.

(2) you don't check that the branch is to the beginning of the loop we care about; it could be a forward branch so a subsequent loop that is aligned

> To make that work properly, we will need block renumbering, after which I can check it based on `bbNum` if it is forward or backward, right?

Reply:
True, no need to change getLoopSize() here. I'm arguing that this code code be the "correct" general case, and getLoopSize can follow later, if necessary.

> To make that work properly, we will need block renumbering, after which I can check it based on bbNum if it is forward or backward, right?

You could save away the block with the ALIGN bit and compare against that when looking for back edges.

However, I think my suggestion is simpler and doesn't require that.